code/config/lib/Loot/Config/Env.hs

```diff
@@ -0,0 +1,361 @@
+
+* APPNAME
+* DB_USERNAME
+* DB_PASSWORD
+-}
+module Loot.Config.Env
+       ( parseEnv
+       , parseEnvPure
+
+         -- * Parsing individual values
+       , EnvValue (..)
+       , Parser
+       , noValue
+       , withPresent
+       , parseStringEnvValue
+       , autoParseEnvValue
+       , parseBoundedNumEnvValue
+
+         -- * Customized parsing
+       , ParseOptions (..)
+       , defaultOptions
+       , parseEnvWith
+       , parseEnvPureWith
+
+         -- * Types
+       , EnvParseError (..)
+       , parseErrorPretty
+       ) where
+
+import Control.Monad.Except (Except, runExcept, throwError)
+import qualified Data.Aeson as Aeson
+import Data.Char (toLower)
+import qualified Data.Fixed as Fixed
+import qualified Data.List.NonEmpty as NE
+import qualified Data.Map as Map
+import qualified Data.Text as T
+import Data.Vinyl (Rec ((:&), RNil))
+import Fmt (Buildable (..), fmt, pretty, (+|), (|+))
+import GHC.TypeLits (KnownSymbol, symbolVal)
+import System.Environment (getEnvironment)
+
+import Loot.Config.Record ((::+), (::-), (:::), (::<), ConfigKind (Partial), ConfigRec,
+                           Item (ItemBranchP, ItemOptionP, ItemSub, ItemSumP), ItemKind,
+                           SumSelection)
+
+-- | A complete description of parsing error.
+data EnvParseError = EnvParseError
+    { errKey     :: Text
+    , errValue   :: Maybe String
+    , errMessage :: Text
+    } deriving (Show, Eq)
+
+instance Buildable EnvParseError where
+    build EnvParseError{..} =
+        -- TODO: can it be that value is secret and should not be displayed?
+        -- In all scenarios which I have encountered if a value have been failed
+        -- to parse, then it was too incorrect and worth throwing away, though
+        -- maybe there are cases when it is mostly not :thinking:
+        -- My idea is to add "isSecure" flag to 'EnvValue' typelass to resolve
+        -- this problem.
+        "Failed to parse environmental variable \
+        \" +| errKey |+ "=" +| maybe "-" build errValue |+ "\
+        \: " +| errMessage |+ ""
+
+-- | Pretty-print a 'EnvParseError'.
+parseErrorPretty :: EnvParseError -> String
+parseErrorPretty = pretty
+
+instance Exception EnvParseError where
+    displayException = fmt . build
+
+-- | Environmental value parser.
+--
+-- Use 'fail' to report parsing errors, and 'noValue' to indicate
+-- that value is left uninitialized.
+newtype Parser a = Parser
+    { runParser :: MaybeT (Except Text) a
+      -- ^ Parsing result, Maybe layer designates result presence, while Except
+      -- layer contains parsing errors.
+    } deriving (Functor, Applicative, Monad)
+
+instance MonadFail Parser where
+    fail = Parser . lift . throwError . fromString
+
+-- | Leave value uninitialized in config.
+noValue :: Parser a
+noValue = Parser mzero
+
+-- | Describes a way to parse an item appearing in config.
+class EnvValue a where
+    -- | Parse a variable value.
+    parseEnvValue :: Maybe String -> Parser a
```

Comment:
So, the way `loot-config` currently works (and I know it is not entirely clear from the docs and from the code structure – that’s why I want to rewrite it from scratch with a different design that emphasises monoids more) is it “automatically wraps” all config options into a `Maybe`, which indicates whether the option was or was not present in the (part of the) config. In this way all options are essentially considered required (so finalising the config will fail if some of them are missing). Since all options are required, it really doesn’t make a lot of sense to allow to “unset” them in parts of configuration.

However, users can “manually” make non-mandatory options, but wrapping them into a `Maybe` themselves. So, instead of `a` you’ll put `Maybe a` in your config declaration, and then `loot-conf` will wrap it in yet another `Maybe` and will work with `Maybe (Maybe a)` – the external maybe is its “internal” thing that shows whether the option was present in at least one config part, and it doesn’t care about the internal `Maybe` – it’s fully transparent for `loot-config` and, let’s say, managed by the user, so they can have a config (part) where they provide the option (they still have to provide the key!) with no value, and it will automatically get parsed as `Nothing` (or, at least, I hope it works this way 😓).

I’m afraid I already lost the line of my thought...

Anyway, the point is that, there is no code in `loot-config` that handles this latter case, it just works naturally (or, at least, is supposed to work naturally) because the `FromJSON` instance for `Maybe a` parses empty string as `Nothing` (again, just to be clear, this `Maybe` is not the one that `loot-config` adds, it comes from the user’s config spec and `loot-config` doesn’t know about it and doesn’t care). So, if you make sure that your `FromEnv` instance for `Maybe a` parses the empty string as `Nothing`, everything will just work™. There is no special case for this in `loot-config` now, and there is not need to have it in your new code.

Reply:
_Holy shit, that’s so long, I hope it is possible to understand what I wrote. (That’s my punishment for not writing a design doc that would explain how `loot-config` actually works.)_

Reply:
Oh yeah, sounds fair. I now see that I wrote some nonsense there, and that scenario I described was supported from the very beginning.

A substantial part of my motivation (which I somehow haven't expressed) was that user may want to declare a special type where the absence of env value is automatically interpreted as some special value (e.g. `MyInt` which is initiated as `0` unless the respective env var is set). But my recent practice shows that in most cases it is better to decouple parsing and default values, so the already supported approaches should be enough for all cases: when there is a static default value, we provide it as `configFromFile <> defaultConfig`, and if the business logic certainly needs to know whether the value was absent to perform some nontrivial computation, having `Maybe a` is enough.

In either case, I find this logic hacky and indeed everything gets simpler if we handle `no value = Nothing` case inside.